### PR TITLE
v2http: client certificate auth via common name

### PR DESCRIPTION
**Pull request by gyuho**
Cherry-picked from https://github.com/coreos/etcd/pull/3916.

### CODE REVIEW

File: etcdserver/api/v2http/client_auth.go [hunk 1]
```diff
@@ -45,26 +82,26 @@ func hasRootAccess(sec auth.Store, r *http.Request) bool {
...
-		return false
-	}
 
-	ok = sec.CheckPassword(rootUser, password)
-	if !ok {
-		plog.Warningf("auth: wrong password for user %s", username)
-		return false
+	var rootUser *auth.User
+	if r.Header.Get("Authorization") == "" {
+		rootUser = userFromClientCertificate(sec, r)
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth.go]: we probably should only enable this when client auth is set: https://github.com/coreos/etcd/blob/master/etcdmain/config.go#L174

Or we trust random client certs.

/cc @gtank @heyitsanthony any opinion?

Comment (heyitsanthony) [inline etcdserver/api/v2http/client_auth.go]: I agree it should ensure the cert's been signed by the client ca

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony Added `ClientTLSCertEnabled` flag to use common name only when the client TLS cert auth is enabled.

PTAL.

Thanks.

File: embed/etcd.go [hunk 2]
```diff
@@ -116,6 +116,7 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 		QuotaBackendBytes:       cfg.QuotaBackendBytes,
 		StrictReconfigCheck:     cfg.StrictReconfigCheck,
 		EnablePprof:             cfg.EnablePprof,
+		ClientTLSCertEnabled:    cfg.ClientTLSInfo.ClientCertAuth,
```
Comment (xiang90) [inline embed/etcd.go]: Probably we should just name this as ClientCertAuthEnabled

File: etcdserver/server.go [hunk 3]
```diff
@@ -208,6 +208,9 @@ type EtcdServer struct {
 	// wg is used to wait for the go routines that depends on the server state
 	// to exit when stopping the server.
 	wg sync.WaitGroup
+
+	// ClientTLSCertEnabled is true when cert has been signed by the client CA.
+	ClientTLSCertEnabled bool
```
Comment (xiang90) [inline etcdserver/server.go]: this is already in the Cfg  at line163, no?

Comment (gyuho) [inline etcdserver/server.go]: Right. Will remove this duplicate flag.

**Comment by gyuho** [comment]
@xiang90 Just addressed. PTAL.

Thanks.

**Comment by gyuho** [comment]
Do not merge yet. Will add e2e tests for this.

**Comment by gyuho** [comment]
@xiang90 Just added e2e tests. Also confirmed manually that it works without username and password when CN is provided:

```
$ ./bin/etcdctl \
  --endpoints=https://localhost:2379 \
  --cert-file=hack/tls-setup/certs/ca.pem \
  --key-file=hack/tls-setup/certs/ca-key.pem \
  --ca-file=hack/tls-setup/certs/ca.pem \
  set foo bar
bar

$ ./bin/etcdctl \
  --endpoints=https://localhost:2379 \
  --cert-file=hack/tls-setup/certs/ca.pem \
  --key-file=hack/tls-setup/certs/ca-key.pem \
  --ca-file=hack/tls-setup/certs/ca.pem \
  set hello world
Error:  110: The request requires user authentication (Insufficient credentials) [0]
```

**Comment by gyuho** [comment]
Also confirmed that if different key given, auth feature correctly rejects the request

```
ctl_v2_test.go:311: failed to write (read /dev/ptmx: input/output error (expected "bar", got ["Error:  110: The request requires user authentication (Insufficient credentials) [0]\r\n"]))
```

**Comment by xiang90** [comment]
lgtm.

**Comment by swsnider** [comment]
I see that this has been assigned the kind/backport label. Will this be backported?

**Comment by gyuho** [comment]
@swsnider Yes this has been backported to https://github.com/coreos/etcd/releases/tag/v3.0.4.

You need etcd v3.0.4+ for this feature.

Thanks.

**Comment by geekofalltrades** [comment]
Is this feature documented anywhere?

**Comment by jesseendahl** [comment]
@geekofalltrades it doesn't appear to be. I just opened a new issue at #8020.

**Pull request merged**